File: src/stores/credentials.ts
import { create } from "zustand";
import {
  credentialAssignProfile,
  credentialCreate,
  credentialDelete,
  credentialList,
  credentialOpenManager,
  credentialReveal,
  credentialSetPin,
  credentialSwitch,
  credentialUpdate,
} from "@/ipc";
import { type AppError, toAppError } from "@/ipc/errors";
import type { Credential } from "@/ipc/types.gen";
import { useActivityStore } from "@/stores/activity";
import { useFeedbackStore } from "@/stores/feedback";

export interface CreateCredentialInput {
  profileId: string | null;
  host: string;
  username: string;
  token: string;
  pin: string | null;
}

export type RevealResult = { ok: true; token: string } | { ok: false; message: string };

interface CredentialsStore {
  items: Credential[];
  loading: boolean;
  error: AppError | null;
  createOpen: boolean;
  editing: Credential | null;
  revealing: Credential | null;
  settingPin: Credential | null;
  setCreateOpen: (open: boolean) => void;
  setEditing: (credential: Credential | null) => void;
  setRevealing: (credential: Credential | null) => void;
  setSettingPin: (credential: Credential | null) => void;
  fetch: () => Promise<void>;
  create: (input: CreateCredentialInput) => Promise<boolean>;
  setPin: (id: string, pin: string) => Promise<boolean>;
  reveal: (id: string, pin: string) => Promise<RevealResult>;
  update: (id: string, username: string, token: string | null) => Promise<boolean>;
  assignProfile: (id: string, profileId: string | null) => Promise<void>;
  switchCredential: (id: string) => Promise<void>;
  remove: (id: string) => Promise<boolean>;
  openManager: () => Promise<void>;
}

export const useCredentialsStore = create<CredentialsStore>((set, get) => ({
  items: [],
  loading: false,
  error: null,
  createOpen: false,
  editing: null,
  revealing: null,
  settingPin: null,

  setCreateOpen: (open) => set({ createOpen: open }),
  setEditing: (credential) => set({ editing: credential }),
  setRevealing: (credential) => set({ revealing: credential }),
  setSettingPin: (credential) => set({ settingPin: credential }),

  fetch: async () => {
    set({ loading: true, error: null });
    try {
      const items = await credentialList();
      set({ items, loading: false });
    } catch (e) {
      set({ error: toAppError(e), loading: false });
    }
  },

  create: async (input) => {
    try {
      await credentialCreate(input.profileId, input.host, input.username, input.token, input.pin);
      await get().fetch();
      useFeedbackStore.getState().toast("Credential created", "success");
      useActivityStore.getState().fetch();
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  setPin: async (id, pin) => {
    try {
      await credentialSetPin(id, pin);
      await get().fetch();
      useFeedbackStore.getState().toast("Reveal PIN set", "success");
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  reveal: async (id, pin) => {
    try {
      const token = await credentialReveal(id, pin);
      return { ok: true, token };
    } catch (e) {
      return { ok: false, message: toAppError(e).message };
    }
  },

  update: async (id, username, token) => {
    try {
      await credentialUpdate(id, username, token);
      await get().fetch();
      useFeedbackStore.getState().toast("Credential updated", "success");
      useActivityStore.getState().fetch();
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  assignProfile: async (id, profileId) => {
    try {
      await credentialAssignProfile(id, profileId);
      await get().fetch();
      useFeedbackStore
        .getState()
        .toast(profileId ? "Credential assigned to profile" : "Credential unassigned", "success");
      useActivityStore.getState().fetch();
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },

  switchCredential: async (id) => {
    try {
      await credentialSwitch(id);
      await get().fetch();
      useFeedbackStore.getState().toast("Credential is now active for its host", "success");
      useActivityStore.getState().fetch();
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },

  remove: async (id) => {
    try {
      await credentialDelete(id);
      await get().fetch();
      useFeedbackStore.getState().toast("Credential deleted", "success");
      useActivityStore.getState().fetch();
      return true;
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
      return false;
    }
  },

  openManager: async () => {
    try {
      await credentialOpenManager();
    } catch (e) {
      useFeedbackStore.getState().toast(toAppError(e).message, "error");
    }
  },
}));

File: src-tauri/src/commands/credentials.rs
use crate::{domain::credential::Credential, error::AppError, state::AppState};
use tauri::State;
use uuid::Uuid;
use zeroize::Zeroizing;

#[tauri::command]
pub(crate) async fn credential_list(
    state: State<'_, AppState>,
) -> Result<Vec<Credential>, AppError> {
    state.credentials.list()
}

#[tauri::command]
pub(crate) async fn credential_create(
    profile_id: Option<Uuid>,
    host: String,
    username: String,
    token: String,
    pin: Option<String>,
    state: State<'_, AppState>,
) -> Result<Credential, AppError> {
    // Own the secret in a zeroizing buffer for its whole (short) lifetime.
    let secret = Zeroizing::new(token);
    let credential = state
        .credentials
        .create(profile_id, &host, username, &secret)?;
    // An optional PIN makes the token revealable later. Apply it as a second
    // step; the credential itself is already safely stored if this is skipped.
    match pin {
        Some(pin) => {
            let pin = Zeroizing::new(pin);
            state.credentials.set_pin(credential.id, &pin)
        }
        None => Ok(credential),
    }
}

/// Set (or replace) a credential's reveal PIN. Enables revealing the token for
/// credentials that were created without one.
#[tauri::command]
pub(crate) async fn credential_set_pin(
    id: Uuid,
    pin: String,
    state: State<'_, AppState>,
) -> Result<Credential, AppError> {
    let pin = Zeroizing::new(pin);
    state.credentials.set_pin(id, &pin)
}

/// Read a credential's token back after verifying its PIN. Returns the raw token
/// to the caller — the single command that ever surfaces secret material.
#[tauri::command]
pub(crate) async fn credential_reveal(
    id: Uuid,
    pin: String,
    state: State<'_, AppState>,
) -> Result<String, AppError> {
    let pin = Zeroizing::new(pin);
    let secret = state.credentials.reveal(id, &pin)?;
    Ok(secret.as_str().to_owned())
}

#[tauri::command]
pub(crate) async fn credential_update(
    id: Uuid,
    username: String,
    token: Option<String>,
    state: State<'_, AppState>,
) -> Result<Credential, AppError> {
    let rotated = token.is_some();
    let secret = token.map(Zeroizing::new);
    let credential = state.credentials.update(id, username, secret.as_ref())?;
    // A rotated token only lands in the backing vault slot. Re-promote it onto
    // the canonical and per-repo path-scoped targets that Git actually reads, so
    // an assigned profile keeps pushing without a manual re-assignment. Best-
    // effort: the token is already saved, so a re-pin hiccup must not fail here.
    if rotated {
        if let Some(profile_id) = credential.profile_id {
            let _ = state.profiles.repin_credentials(profile_id);
        }
    }
    Ok(credential)
}

#[tauri::command]
pub(crate) async fn credential_assign_profile(
    id: Uuid,
    profile_id: Option<Uuid>,
    state: State<'_, AppState>,
) -> Result<Credential, AppError> {
    state.credentials.assign_profile(id, profile_id)
}

#[tauri::command]
pub(crate) async fn credential_switch(
    id: Uuid,
    state: State<'_, AppState>,
) -> Result<Credential, AppError> {
    state.credentials.switch(id)
}

#[tauri::command]
pub(crate) async fn credential_delete(
    id: Uuid,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    state.credentials.delete(id)
}

/// Open the native Windows Credential Manager UI. Windows-only — no reliable,
/// universal way to launch a credential manager exists on Linux/macOS, so this
/// returns `Unsupported` there. The frontend gates the action on
/// `platform_capabilities().native_credential_manager`, so this error is a
/// backend safety net, not something a user reaches by clicking a visible button.
#[tauri::command]
pub(crate) async fn credential_open_manager() -> Result<(), AppError> {
    #[cfg(windows)]
    {
        std::process::Command::new("rundll32.exe")
            .args(["keymgr.dll,KRShowKeyMgr"])
            .spawn()
            .map_err(|e| AppError::Io(e.to_string()))?;
        Ok(())
    }
    #[cfg(not(windows))]
    {
        Err(AppError::Unsupported(
            "the Windows Credential Manager is only available on Windows".into(),
        ))
    }
}
